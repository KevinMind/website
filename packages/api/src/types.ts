import { z } from "zod";

// Priority enum based on Django model
export const TodoPriority = z.enum(["low", "medium", "high"]);
export type TodoPriority = z.infer<typeof TodoPriority>;

// Todo item schema based on Django serializer
export const TodoItemSchema = z.object({
  id: z.number(),
  title: z.string(),
  description: z.string().nullable(),
  completed: z.boolean(),
  priority: TodoPriority,
  due_date: z.string().datetime().nullable(),
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
  is_overdue: z.boolean(),
});

export type TodoItem = z.infer<typeof TodoItemSchema>;

// Create todo schema (what we send to create a new todo)
export const CreateTodoSchema = z.object({
  title: z.string().min(1, "Title cannot be empty"),
  description: z.string().optional(),
  priority: TodoPriority.optional().default("medium"),
  due_date: z.string().datetime().optional(),
});

export type CreateTodo = z.infer<typeof CreateTodoSchema>;

// Update todo schema (what we send to update a todo)
export const UpdateTodoSchema = z.object({
  title: z.string().min(1, "Title cannot be empty").optional(),
  description: z.string().optional(),
  completed: z.boolean().optional(),
  priority: TodoPriority.optional(),
  due_date: z.string().datetime().optional(),
});

export type UpdateTodo = z.infer<typeof UpdateTodoSchema>;

// Stats schema for the stats endpoint
export const TodoStatsSchema = z.object({
  total: z.number(),
  completed: z.number(),
  incomplete: z.number(),
  overdue: z.number(),
  by_priority: z.object({
    low: z.number(),
    medium: z.number(),
    high: z.number(),
  }),
});

export type TodoStats = z.infer<typeof TodoStatsSchema>;

// API response schemas
export const PaginatedTodosSchema = z.object({
  count: z.number(),
  next: z.string().url().nullable(),
  previous: z.string().url().nullable(),
  results: z.array(TodoItemSchema),
});

export type PaginatedTodos = z.infer<typeof PaginatedTodosSchema>;

// API Error schema
export const ApiErrorSchema = z.object({
  detail: z.string().optional(),
  errors: z.record(z.array(z.string())).optional(),
});

export type ApiError = z.infer<typeof ApiErrorSchema>;

// Bulk operation response schemas
export const BulkOperationResponseSchema = z.object({
  updated_count: z.number().optional(),
  deleted_count: z.number().optional(),
  message: z.string(),
});

export type BulkOperationResponse = z.infer<typeof BulkOperationResponseSchema>;