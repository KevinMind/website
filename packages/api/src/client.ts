import {
  TodoItem,
  TodoItemSchema,
  CreateTodo,
  CreateTodoSchema,
  UpdateTodo,
  UpdateTodoSchema,
  TodoStats,
  TodoStatsSchema,
  PaginatedTodos,
  PaginatedTodosSchema,
  BulkOperationResponse,
  BulkOperationResponseSchema,
  ApiError as ApiErrorType,
  ApiErrorSchema,
} from "./types";

export class TodoApiError extends Error {
  constructor(
    message: string,
    public status: number,
    public errors?: Record<string, string[]>
  ) {
    super(message);
    this.name = "ApiError";
  }
}

export interface TodoApiConfig {
  baseUrl?: string;
  defaultHeaders?: Record<string, string>;
}

export class TodoApi {
  private baseUrl: string;
  private defaultHeaders: Record<string, string>;

  constructor(config: TodoApiConfig = {}) {
    this.baseUrl = config.baseUrl || "http://localhost:8000";
    this.defaultHeaders = {
      "Content-Type": "application/json",
      ...config.defaultHeaders,
    };
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {},
    schema?: any
  ): Promise<T> {
    const url = `${this.baseUrl}/todo/items${endpoint}`;
    
    const response = await fetch(url, {
      ...options,
      headers: {
        ...this.defaultHeaders,
        ...options.headers,
      },
    });

    if (!response.ok) {
      let errorData: ApiErrorType | undefined;
      try {
        const errorJson = await response.json();
        errorData = ApiErrorSchema.parse(errorJson);
      } catch {
        // If we can't parse the error, create a generic one
      }

      throw new TodoApiError(
        errorData?.detail || `HTTP ${response.status}: ${response.statusText}`,
        response.status,
        errorData?.errors
      );
    }

    const data = await response.json();
    
    if (schema) {
      return schema.parse(data);
    }
    
    return data;
  }

  // List todos with optional filtering
  async listTodos(params?: {
    completed?: boolean;
    priority?: "low" | "medium" | "high";
    search?: string;
    ordering?: string;
    page?: number;
  }): Promise<PaginatedTodos> {
    const searchParams = new URLSearchParams();
    
    if (params?.completed !== undefined) {
      searchParams.set("completed", String(params.completed));
    }
    if (params?.priority) {
      searchParams.set("priority", params.priority);
    }
    if (params?.search) {
      searchParams.set("search", params.search);
    }
    if (params?.ordering) {
      searchParams.set("ordering", params.ordering);
    }
    if (params?.page) {
      searchParams.set("page", String(params.page));
    }

    const queryString = searchParams.toString();
    const endpoint = queryString ? `/?${queryString}` : "/";
    
    return this.request(endpoint, { method: "GET" }, PaginatedTodosSchema);
  }

  // Get a single todo by ID
  async getTodo(id: number): Promise<TodoItem> {
    return this.request(`/${id}/`, { method: "GET" }, TodoItemSchema);
  }

  // Create a new todo
  async createTodo(todo: CreateTodo): Promise<TodoItem> {
    const validatedTodo = CreateTodoSchema.parse(todo);
    
    return this.request(
      "/",
      {
        method: "POST",
        body: JSON.stringify(validatedTodo),
      },
      TodoItemSchema
    );
  }

  // Update a todo
  async updateTodo(id: number, updates: UpdateTodo): Promise<TodoItem> {
    const validatedUpdates = UpdateTodoSchema.parse(updates);
    
    return this.request(
      `/${id}/`,
      {
        method: "PATCH",
        body: JSON.stringify(validatedUpdates),
      },
      TodoItemSchema
    );
  }

  // Delete a todo
  async deleteTodo(id: number): Promise<void> {
    await this.request(`/${id}/`, { method: "DELETE" });
  }

  // Mark a todo as completed
  async completeTodo(id: number): Promise<TodoItem> {
    return this.request(
      `/${id}/complete/`,
      { method: "POST" },
      TodoItemSchema
    );
  }

  // Mark a todo as incomplete
  async uncompleteTodo(id: number): Promise<TodoItem> {
    return this.request(
      `/${id}/uncomplete/`,
      { method: "POST" },
      TodoItemSchema
    );
  }

  // Get todo statistics
  async getStats(): Promise<TodoStats> {
    return this.request("/stats/", { method: "GET" }, TodoStatsSchema);
  }

  // Complete all todos
  async completeAll(): Promise<BulkOperationResponse> {
    return this.request(
      "/complete_all/",
      { method: "POST" },
      BulkOperationResponseSchema
    );
  }

  // Clear completed todos
  async clearCompleted(): Promise<BulkOperationResponse> {
    return this.request(
      "/clear_completed/",
      { method: "DELETE" },
      BulkOperationResponseSchema
    );
  }
}

// Default instance for easy importing
export const todoApi = new TodoApi();