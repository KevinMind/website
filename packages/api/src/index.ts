// Export types
export * from "./types";

// Export client (excluding conflicting ApiError which is now TodoApiError)
export { TodoApi, TodoApiError, todoApi, type TodoApiConfig } from "./client";

// Export default instance for convenience
export { todoApi as default } from "./client";