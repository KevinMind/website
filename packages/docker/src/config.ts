import { path } from "zx";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const root = path.join(__dirname, "..", "..", "..");

export const config = {
  root,
  env: path.join(root, ".env"),
  composeFile: path.join(root, "docker-compose.yml"),
} as const;
