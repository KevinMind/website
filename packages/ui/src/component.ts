import { LitElement, html, css } from 'lit';
import { customElement } from 'lit/decorators.js';

@customElement('ui-component')
export class Component extends LitElement {
  static override styles = css`
    :host {
      display: block;
      padding: 16px;
      border: 1px solid #ccc;
      border-radius: 4px;
    }
  `;

  override render() {
    return html`
      <div>Component</div>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    'ui-component': Component;
  }
}

export default Component;
