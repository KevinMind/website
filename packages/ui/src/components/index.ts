// Export all components
export { Button, buttonVariants, type ButtonProps } from './button';