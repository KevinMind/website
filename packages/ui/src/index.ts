const ui = 'foo';

export default ui;
