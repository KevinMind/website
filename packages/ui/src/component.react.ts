import React from "react";

import { createComponent } from "@lit/react";

import { MoreComponent } from "./component";

export const MoreComponentReact = createComponent({
  tagName: "more-component",
  elementClass: MoreComponent,

  react: React,

  events: {
    onactivate: "activate",
    onchange: "change",
  },
});
