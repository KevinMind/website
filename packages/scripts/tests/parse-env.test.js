import { describe, it, expect } from "vitest";
import { tmpfile, dotenv } from "zx";

import { parseEnv } from "../src/parse-env";

const DOCKER_DIGEST =
  "1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef";

function createTmpEnv(env = {}) {
  const str = dotenv.stringify(env);
  const tmpPath = tmpfile("env", str);
  return tmpPath;
}

describe("parseEnv", () => {
  it("should parse the env", () => {
    const env = parseEnv(createTmpEnv());
    expect(env).toMatchSnapshot();
  });

  it("should override from file", () => {
    const env = parseEnv(
      createTmpEnv({
        NODE_ENV: "production",
      }),
      {},
    );
    expect(env.NODE_ENV).toBe("production");
  });

  it("should override from process", () => {
    const env = parseEnv(
      createTmpEnv({
        NODE_ENV: "production",
      }),
      {
        NODE_ENV: "development",
      },
    );
    expect(env.NODE_ENV).toBe("development");
  });

  it("rejects invalid NODE_ENV", () => {
    expect(() => parseEnv(undefined, { NODE_ENV: "invalid" })).toThrow();
  });

  it("rejects invalid DOCKER_TARGET", () => {
    expect(() => parseEnv(undefined, { DOCKER_TARGET: "invalid" })).toThrow();
  });

  it("should use provided DOCKER_TAG if present", () => {
    const env = parseEnv(undefined, {
      DOCKER_TAG: "docker.io/custom/image:1.0.0",
    });
    expect(env.DOCKER_TAG).toBe("docker.io/custom/image:1.0.0");
    expect(env.DOCKER_REGISTRY).toBe("docker.io");
    expect(env.DOCKER_IMAGE).toBe("custom/image");
    expect(env.DOCKER_VERSION).toBe("1.0.0");
  });

  it("should not construct tag from components if DOCKER_TAG not present", () => {
    const env = parseEnv(undefined, {
      DOCKER_REGISTRY: "quay.io",
      DOCKER_IMAGE: "custom/image",
      DOCKER_VERSION: "2.0.0",
    });
    expect(env.DOCKER_TAG).toBe("docker.io/mozilla/addons-server:local");
  });

  it("should use base tag when no docker config provided", () => {
    const env = parseEnv(undefined, {});
    expect(env.DOCKER_TAG).toBe("docker.io/mozilla/addons-server:local");
  });

  it("should force production target for non-local image", () => {
    const env = parseEnv(undefined, {
      DOCKER_TAG: "docker.io/mozilla/addons-server:1.0.0",
      DOCKER_TARGET: "development",
    });
    expect(env.DOCKER_TARGET).toBe("production");
  });

  it.each(["development", "production"])(
    "should allow either target for local image",
    (target) => {
      const env = parseEnv(undefined, {
        DOCKER_TAG: "docker.io/mozilla/addons-server:local",
        DOCKER_TARGET: target,
      });
      expect(env.DOCKER_TARGET).toBe(target);
    },
  );

  it("should throw error when digest provided without version", () => {
    expect(() =>
      parseEnv(undefined, {
        DOCKER_TAG: `@sha256:${DOCKER_DIGEST}`,
      }),
    ).toThrow("Invalid image tag: ");
  });

  it("should handle version-only input correctly", () => {
    const env = parseEnv(undefined, {
      DOCKER_TAG: "3.0.0",
    });
    expect(env.DOCKER_TAG).toBe("docker.io/mozilla/addons-server:3.0.0");
  });

  it("should handle version and digest input correctly", () => {
    const env = parseEnv(undefined, {
      DOCKER_TAG: `1.0.0@sha256:${DOCKER_DIGEST}`,
    });
    expect(env.DOCKER_TAG).toBe(
      `docker.io/mozilla/addons-server:1.0.0@sha256:${DOCKER_DIGEST}`,
    );
  });

  it("should handle image-only input correctly", () => {
    const env = parseEnv(undefined, {
      DOCKER_TAG: "custom/image:1.0.0",
    });
    expect(env.DOCKER_TAG).toBe("docker.io/custom/image:1.0.0");
  });
});
