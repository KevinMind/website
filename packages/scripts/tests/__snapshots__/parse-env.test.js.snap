// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseEnv > should parse the env 1`] = `
{
  "DOCKER_DIGEST": "",
  "DOCKER_IMAGE": "mozilla/addons-server",
  "DOCKER_REGISTRY": "",
  "DOCKER_TAG": "mozilla/addons-server:local",
  "DOCKER_TARGET": "development",
  "DOCKER_VERSION": "local",
  "NODE_ENV": "development",
}
`;
